src/lib.rs

```diff
@@ -59,6 +66,49 @@ pub struct Pointer<T> {
     pending: PendingRef,
 }
 
+impl<T> Pointer<T> {
+    /// Creates a new `WeakPointer` to this component.
+    pub fn downgrade(&self) -> WeakPointer<T> {
+        let epoch = self.target.read().unwrap().epoch[self.index];
+        WeakPointer {
+            index: self.index,
+            epoch: epoch,
+            target: Arc::downgrade(&self.target),
+            pending: Arc::downgrade(&self.pending),
+        }
+    }
+}
+
+/// Weak variant of `Pointer`.
+pub struct WeakPointer<T> {
+    index: usize,
+    epoch: Epoch,
+    target: WeakStorageRef<T>,
+    pending: WeakPendingRef,
+}
+
+impl<T> WeakPointer<T> {
+    /// Upgrades the `WeakPointer` to an `Pointer`, if possible.
+    /// Returns `None` if the strong count has reached zero and the inner value was destroyed.
+    pub fn upgrade(&self) -> Option<Pointer<T>> {
+        match self.target.upgrade().and(self.pending.upgrade()) {
+            None => None,
+            Some(arc) => {
+                let target = self.target.upgrade().unwrap();
+                if target.read().unwrap().meta[self.index] == 0 ||
```

Comment:
If possible, I'd like the pointer to not even be able to lock the storage. The fact it's currently possible is purely accidental.

> access_weak and access_weak_mut which will return Result<Pointer<T>>.
that would be inconsistent with our current `access` semantics

> But we still need regular upgrade method.

Do we? That use-case is not obvious to me.